src/rest/BackOffRequest.ts

```diff
@@ -0,0 +1,57 @@
+import { ISuccessResponse } from './EndpointCaller';
+import { backOff as backOffModule } from '../misc/BackOff';
+
+let backOff = backOffModule;
+
+export function setBackOffModule(newModule) {
+  backOff = newModule;
+}
+
+type Request = () => Promise<ISuccessResponse<{}>>;
+
+export interface IBackOffRequest {
+  fn: Request;
+  retry?: (e, attemptNumber: number) => boolean;
+}
+
+export class BackOffRequest {
+  private static queue: Request[] = [];
+  private static clearingQueue = false;
+
+  public static enqueue(request: IBackOffRequest): Promise<ISuccessResponse<{}>> {
```

Comment:
I'd change the definition to use the generics you are "semi" using.

```
type Request<T> = () => Promise<ISuccessResponse<T>>;

export interface IBackOffRequest<T> {
  fn: Request<T>;
  retry?: (e, attemptNumber: number) => boolean;
}

enqueue<T>(request: IBackOffRequest<T>): Promise<ISuccessResponse<T>>
```
That kinda stuff, so you would not have to cast when you use it inside SearchEndpoint : 

```
const backOffRequest = { fn: request, retry: this.retryIf429Error };
const response = await BackOffRequest.enqueue<T>(backOffRequest);
return response.data;
```